src/tests/roc_fec/target_openfec/test_writer_reader.cpp

```diff
@@ -797,5 +755,186 @@ TEST(writer_reader, delayed_packets) {
     }
 }
 
+TEST(writer_reader, fill_sn_sbn) {
+    OFEncoder encoder(config, FECPayloadSize, allocator);
+    OFDecoder decoder(config, FECPayloadSize, buffer_pool, allocator);
+
+    CHECK(encoder.valid());
+    CHECK(decoder.valid());
+
+    PacketDispatcher dispatcher;
+
+    Writer writer(config, FECPayloadSize, encoder, dispatcher, source_composer,
+                  repair_composer, packet_pool, buffer_pool, allocator);
+
+    Reader reader(config, decoder, dispatcher.source_reader(),
+                  dispatcher.repair_reader(), rtp_parser, packet_pool, allocator);
+
+    CHECK(writer.valid());
+    CHECK(reader.valid());
+
+    // Send first block.
+    fill_all_packets(0);
+
+    for (size_t n = 0; n < NumSourcePackets; ++n) {
+        writer.write(source_packets[n]);
+    }
+
+    // Receive first block.
+    dispatcher.release_all();
+
+    packet::PacketPtr first_packet = reader.read();
+    CHECK(first_packet);
+    CHECK(first_packet->fec()->repair_symbol_id == 0);
+
+    const packet::blknum_t sbn = first_packet->fec()->source_block_number;
+
+    for (size_t n = 1; n < NumSourcePackets; ++n) {
+        packet::PacketPtr p = reader.read();
+        CHECK(p);
+        CHECK(p->fec()->source_block_number == sbn);
+        CHECK(p->fec()->repair_symbol_id == n);
```

Comment:
It would be great to add two more checks to this test or maybe to a separate test:

* check source_block_length
* check source_block_number and repair_symbol_id of repair packets as well (currently we check only source packets)

Reply:
I added a separate test.

Reply:
I'm looking at `encode_sbn` and `encode_sbn_validate_repair_packets` tests and I think it's not quite correct to use reader there. We want to test that the *writer* correctly fills sbn, sbl, and esi, right? However in current tests we can't be sure that these fields were filled by writer, not by reader. Note that `encode_source_id_and_seqnum` test doesn't use reader. I suggest to remove reader usage from these tests and also I think we can merge them because they test the same feature and look very similarly.

On the other hand, `encode_sbn_outdated_block` and `encode_sbn_repaired_block` actually test writer + reader cooperation, and not the encoding process itself. I think we should rename them to something like `drop_outdated_block` and `repaired_block_numbering`, respectively, and move back to the end of the file.

Reply:
Fixed